### PR TITLE
Add support to conv-relu6 as supergroup

### LINKED ISSUE #672: Add support to identify [Conv - Relu6] as supergroup
- Add support for conv -> relu6 supergroup in quantsim config json file

From a complexity perspective, this ticket is at an easy level.

### PR DESCRIPTION
**Pull request by mohanksriram**
#672 
Signed-off-by: mohanksriram <mohanksriram@gmail.com>

### CODE REVIEW

**Comment by mohanksriram** [comment]
Hi @quic-klhsieh, Thank you for your detailed description of how to approach the issue. I have a few questions for you.

1. I get the following error in my local build inside a docker container:

    cmake_pybind_error

    Have you faced a similar issue before?
    I'm following the instructions [here](https://github.com/quic/aimet/blob/develop/packaging/docker_install.md#docker-information).

2. In order to test both pytorch and tensorflow quantsim_config locally should I create four separate containers with ```AIMET_VARIANT``` set to ```"torch-gpu"```, ```"torch-cpu"```, ```"tf-gpu"```, and ```"tf-cpu"``` respectively.

2. I added conv-relu6 to the supergroup op-list. I also included a torch test case. But the Jenkins test seems to fail with ```TypeError: 'NoneType' object is not iterable```. Should I include ```Relu6``` elsewhere?

**Comment by quic-klhsieh** [comment]
Hi @mohanksriram, answers to your questions as follows:
1. Are you using one of our dockerfiles to run the docker image? In your attached screenshot, it looks like you are running python 3.8, which is a newer version of python than we normally work with (3.6). Can you provide a bit more information the steps you are taking to start the docker and get to the error? We advise you to switch to using python 3.6 when developing AIMET.

2. Correct, if your computer has a compatible gpu, you can develop and test pytorch side code while on torch-gpu, and then separately start up the tf-gpu docker and develop and test tensorflow side code. If you don't have a gpu, you can do the same for torch-cpu and tf-cpu. All four variants will be tested in the Jenkins pipeline.

3. The configuration file is designed to be used by both pytorch as well as tensorflow code, so we have tried to use a common naming convention of onnx operator types in the config file. Relu6 is not an operator in onnx, so you should look to instead use the corresponding operator name (which I believe is Clip, but not 100% sure. Worth double checking.)  Additionally, for pytorch and tensorflow code to parse the onnx type, there is a translation step involving mappings when pytorch and tf quantsim config code is parsing the config file. You may need to add mappings for Relu6 for either framework so that quantsim config code is able to parse the new supergroup type. Please refer to aimet_torch/quantsim_config/quantsim_config.py and aimet_tensorflow/quantsim_config/quantsim_config.py for more details.

**Comment by mohanksriram** [comment]
Hi @quic-klhsieh , please find the follow-up questions below:
1. Yes, I'm using ./buildntest.sh script to choose and run the appropriate docker container. Within the container, I'm calling the cmake .. command. The default python environment on my ubuntu 20.04 machine is 3.8. I might have to modify the docker file to create a conda environment within the docker container that runs python 3.6. What are your thoughts?
3. I've made the change to include "Clip".  I also modified torch TinyModel to include a Relu6 activation at ```model.relu2``` but this change seems to fail an already existing test.

File: TrainingExtensions/tensorflow/src/python/aimet_tensorflow/utils/common.py [hunk 1]
```diff
@@ -55,6 +55,7 @@
     [["MaxPool"], ["MaxPool"]],
     [["AveragePool"], ["AvgPool"]],
     [["Relu"], ["Relu", "Relu6"]],
```
Comment (quic-klhsieh) [inline TrainingExtensions/tensorflow/src/python/aimet_tensorflow/utils/common.py]: Please also remove Relu6 from this line, I believe this is not the correct mapping and the one you have added is what we should have.

**Comment by quic-klhsieh** [comment]
Okay, for docker and installation issues, let me defer to @quic-bharathr who would be able to provide better advice.

The TinyModel is a toy model we have defined which we are using in multiple unit tests, so one of them is probably not expecting relu to be changed to relu6.

For testing purposes, we do reuse toy models when convenient, but in creating new tests if we find we need a change in the model definition that will affect other tests, it is easier to just define another model in test_models.py. Feel free to do this as well.  You can also change the model being used in the unit test and update the test (as long as the unit test still tests everything it had been testing), or define another unit test altogether to test relu6.

**Comment by mohanksriram** [comment]
Thanks, I've made the necessary changes for torch unit tests. I'm not very familiar with tensorflow. I'll need help to modify the existing unit tests to include relu6 layer.

**Comment by quic-klhsieh** [comment]
Sounds good. Regarding tensorflow unit tests, there is a similar test_parse_config_file_supergroups() under tensorflow/test/python/test_quantism_config.py you can modify. Similar to torch side, the model being used may be used in many tests so you can define another one that uses tf.nn.relu6 instead of tf.nn.relu for one of the layers within.

The logic for checking whether a quantizer has been enabled or disabled involves checking the value of a particular tensor in the quantization op which we have inserted as part of quantsim init. In TF 1.15, this involves looking up a tensor with a known name using sim.session.graph.get_tensor_by_name, and then later querying it with sim.session.run.

The list of strings under activation_quantizers represent names for all the quantization ops we have inserted, a subset of which (5 of them) will become disabled due to config settings. We check the enabled/disabled status by evaluating the op mode tensor for each quantizer and checking if the mode is passThrough (disabled) or updateStats (enabled). For 4 out of 5 of the disabled tensors, they correspond to supergroups starting with Convs or Add (the 5th one is input_1_quantized which is disabled because we disable input quantization by default), so we check the quantizer in between the supergroup.

Quantizer ops are named using the name of the preceding tensorflow op, so the quantizer between Conv and BN would be called Conv/BiasAdd_quantized if the Conv had a bias, since the BiasAdd operation is the last operation in the Conv layer. Then the op mode tensor for the quantizer has the name of the quantizer with '_op_mode:0' appended to the name.

To come up with this test case, you need a way to get the names of tf ops in order to know which names to check. This can be done by printing session.graph.get_operations() to get a list of tf ops in the graph.

**Comment by quic-bharathr** [comment]
Hi @mohanksriram
Sorry for the delayed response. In which location are your python packages installed. If they are in a non-standard path, could you please modify this line accordingly?
https://github.com/quic/aimet/blob/develop/CMakeLists.txt#L102

If you use the buildntest.sh, it should automatically build and spin up a docker based on Ubuntu 18 and python 3.6. Just to be safe, can you please replace python and python3-dev in these locations with python3.6 and python3.6-dev respectively?
https://github.com/quic/aimet/blob/develop/Jenkins/Dockerfile.tf-gpu#L71
https://github.com/quic/aimet/blob/develop/Jenkins/Dockerfile.torch-gpu#L68

We currently do not support conda.

PS: You can enter the docker container in interactive mode using the "-i" option with buildntest.sh to check your versions and packages, etc. Or build and start the container manually as follows:
https://github.com/quic/aimet/blob/develop/packaging/docker_install.md#start-docker-container-manually

**Comment by quic-ssiddego** [comment]
Hi @mohanksriram , cc @quic-klhsieh   Why is this marked as a draft PR?  If the changes look fine, please put this up for review and merge the updates. Thank you.

**Comment by quic-klhsieh** [comment]
That works as well. @mohanksriram, your changes for pytorch work so we can merge that in as one PR, and work on Tensorflow side unit tests as another PR.

**Comment by mohanksriram** [comment]
Hi @quic-bharathr, Thanks for your suggestions.

I modified the docker file to use python3.6 as you suggested.

I'm facing installation issues only when I run ```buildntest.sh``` in the interactive mode with "-i" option (python3.8 interpreter is picked). But when I run the container in the non-interactive mode, python3.6 gets used and I do not face any issues.

**Review by quic-klhsieh — APPROVED** [review]

**Pull request merged**